oak_core/src/timer.rs

```diff
@@ -0,0 +1,191 @@
+/// Measures the number of clock cycles between `new()` and `elapsed()`. This measurement only makes
+/// sense if the process is running on the same CPU, as different CPUs will have different tick
+/// counter values.
+pub struct Timer {
+    start: u64,
+}
+
+impl Timer {
+    /// Constructs a new timer, recording the current tick counter value.
+    pub fn new() -> Self {
+        Self { start: rdtsc() }
+    }
+
+    /// Returns the approximate number of clock cycles it took to execute `func`.
+    pub fn timed<F>(func: F) -> u64
+    where
+        F: FnOnce(),
+    {
+        let timer = Timer::new();
+        func();
+        timer.elapsed()
+    }
+
+    /// Returns the approximate number of clock cycles elapsed since the construction of the
+    /// `Timer`.
+    pub fn elapsed(&self) -> u64 {
+        let stop = rdtsc();
+        stop - self.start
+    }
+}
+
+/// Basic histogram to compute percentiles over some measurement.
+/// N is the number of elements kept in memory; it must be larger than 0.
+/// The array is allocated statically; Histogram does not use heap allocation.
+#[derive(Debug, PartialEq)]
+pub struct Histogram<const N: usize> {
+    data: [u64; N],
+    samples: usize,
+}
+
+impl<const N: usize> Histogram<N> {
+    pub fn new() -> Option<Self> {
+        if N == 0 {
+            None
+        } else {
+            Some(Self {
+                data: [0; N],
+                samples: 0,
+            })
+        }
+    }
+
+    /// Records a new data point value.
+    ///
+    /// If the histogram has already been filled, it will probabilistically replace one of the
+    /// existing entries based on reservoir sampling.
+    pub fn record(&mut self, value: u64) {
+        // Reservoir sampling. See https://en.wikipedia.org/wiki/Reservoir_sampling for the description.
+        let j = if self.samples < self.data.len() {
+            self.samples
+        } else {
+            // generate a random number in [0..samples)
+            let mut buf = [0u8; 8];
+            getrandom::getrandom(&mut buf[..]).unwrap();
+            // This will introduce a small bias, but it should be small enough for our use case.
+            usize::from_ne_bytes(buf) % self.samples
+        };
+
+        // if j < N then replace data[j]
+        if let Some(elem) = self.data.get_mut(j) {
+            *elem = value;
+        }
+
+        self.samples += 1;
+    }
+
+    /// Gets the n-th percentile. n needs to be between 0 and 100, otherwise None is returned; also,
+    /// the histogram must have at least one recorded data point.
+    pub fn percentile(&self, n: u8) -> Option<u64> {
+        if n > 100 || self.samples == 0 {
+            return None;
+        }
+        let mut array = self.data;
+        // If we've had more samples, cap the size of the slice at N.
+        let size = core::cmp::min(N, self.samples);
+        let slice = &mut array[..size];
+        slice.sort_unstable();
+        let index = ((size - 1) * n as usize) / 100;
```

Comment:
I am a bit unsure about this. It feels like this might introduce some bias, but I don't immdiately have a better suggestion. Perhaps you could add a test or two to cover edge cases, specificallly with large sizes.

Reply:
I added a test with a store of 1000, and it seems to work, at least when filled with predictable data.